opp_gui/src/widgets/tool_path_parameters_editor_widget.cpp

```diff
@@ -165,22 +272,31 @@ void ToolPathParametersEditorWidget::onGenerateToolPathsComplete(
   if (state != actionlib::SimpleClientGoalState::SUCCEEDED)
   {
     std::string message = "Action '" + GENERATE_TOOLPATHS_ACTION + "' failed to succeed";
-    QMessageBox::warning(this, "Tool Path Planning Error", QString(message.c_str()));
+    emit QWarningBox(message);
   }
   else
   {
     if (!res->success || !res->tool_path_validities[0])
     {
-      QMessageBox::warning(this, "Tool Path Planning Error", "Tool path generation failed");
+      emit QWarningBox("Tool path generation failed");
+    }
+    else if (res->tool_paths[0].paths.size() < 1 || res->tool_paths[0].paths[0].segments.size() < 1)
+    {
+      emit QWarningBox("Tool path generation succeeded, but not paths returned");
     }
     else
     {
-      ROS_INFO_STREAM("Successfully generated tool path");
-
+      ROS_INFO_STREAM("Successfully generated tool path returned from action");
       opp_msgs::ToolPath tp;
       tp.header.stamp = ros::Time::now();
       tp.process_type.val = qvariant_cast<opp_msgs::ProcessType::_val_type>(ui_->combo_box_process_type->currentData());
-      tp.paths = res->tool_paths[0].paths[0].segments;  // TODO, do something smart with the array of arrays issue here
+      for (size_t i = 0; i < res->tool_paths[0].paths.size(); i++)
+      {
+        for (size_t j = 0; j < res->tool_paths[0].paths[i].segments.size(); j++)
+        {
+          tp.paths.push_back(res->tool_paths[0].paths[i].segments[j]);
+        }
+      }
```

Comment:
We only send one mesh to be processed, so tool_path.size() should always be 1. Noether supports sending multiple meshes and getting multiple tool_paths[N]. Each path[] in tool_paths[0] can have multiple segments. each path[i] represents a plane slice through the mesh. However it may not be continuous, so we get multiple segments.